test/test_puma_server.rb

```diff
@@ -104,7 +104,64 @@ def test_respect_x_forwarded_proto
 
     req = Net::HTTP::Get.new("/")
     req['HOST'] = "example.com"
-    req['X_FORWARDED_PROTO'] = "https"
+    req['X_FORWARDED_PROTO'] = "https,http"
+
+    res = Net::HTTP.start @host, @server.connected_port do |http|
+      http.request(req)
+    end
+
+    assert_equal "443", res.body
+  end
+
+  def test_respect_x_forwarded_ssl_on
+    @server.app = proc do |env|
+      [200, {}, [env['SERVER_PORT']]]
+    end
+
+    @server.add_tcp_listener @host, @port
+    @server.run
+
+    req = Net::HTTP::Get.new("/")
+    req['HOST'] = "example.com"
+    req['X_FORWARDED_SSL'] = "on"
+
+    res = Net::HTTP.start @host, @server.connected_port do |http|
+      http.request(req)
+    end
+
+    assert_equal "443", res.body
+  end
+
+  def test_respect_x_forwarded_ssl_off
+    @server.app = proc do |env|
+      [200, {}, [env['SERVER_PORT']]]
+    end
+
+    @server.add_tcp_listener @host, @port
+    @server.run
+
+    req = Net::HTTP::Get.new("/")
+    req['HOST'] = "example.com"
+    req['X_FORWARDED_SSL'] = "off"
+
+    res = Net::HTTP.start @host, @server.connected_port do |http|
+      http.request(req)
+    end
+
+    assert_equal "80", res.body
```

Comment:
I wonder - why not just test the actual methods modified rather than test res.body? This test is testing a lot more than it has to.